File: src/main/java/com/wooteco/wiki/document/repository/DocumentRepository.java
package com.wooteco.wiki.document.repository;

import com.wooteco.wiki.document.domain.Document;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.UUID;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

public interface DocumentRepository extends JpaRepository<Document, Long> {

    Optional<Document> findByTitle(String title);

    Boolean existsByTitle(String title);

    List<Document> findAllByTitleStartingWithOrderByTitle(String keyWord);

    Optional<Document> findByUuid(UUID uuid);

    @Query("SELECT d.uuid FROM Document d WHERE d.title = :title")
    Optional<UUID> findUuidByTitle(@Param("title") String title);

    @Query("SELECT d.id FROM Document d WHERE d.uuid = :uuid")
    Optional<Long> findIdByUuid(@Param("uuid") UUID documentUuid);

    List<Document> findAllByUuidIn(Set<UUID> uuids);
}

File: src/main/java/com/wooteco/wiki/document/repository/DocumentRepository.kt
package com.wooteco.wiki.document.repository

import com.wooteco.wiki.document.domain.Document
import com.wooteco.wiki.document.domain.Title
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import java.util.*

interface DocumentRepository : JpaRepository<Document, Long> {

    fun findByTitle(title: String): Optional<Document>

    fun existsByTitle(title: String): Boolean

    fun findAllByOrderByGenerateTimeDesc(): List<Document>

    fun findAllByTitleStartingWith(keyWord: String): List<Title>

    fun findByUuid(uuid: UUID): Optional<Document>

    @Query("SELECT d.uuid FROM Document d WHERE d.title = :title")
    fun findUUidByTitle(@Param("title") title: String): Optional<UUID>
}
